backbone.js

```diff
@@ -224,27 +158,99 @@
   // in terms of the existing API.
-  var eventsApi = function(obj, action, name, rest) {
-    if (!name) return true;
-
-    // Handle event maps.
-    if (typeof name === 'object') {
-      for (var key in name) {
-        obj[action].apply(obj, [key, name[key]].concat(rest));
+  var eventsApi = function(api, events, name, callback, context, ctx) {
+    var i = 0, names, length;
+    if (name && typeof name === 'object') {
+      // Handle event maps.
+      for (names = _.keys(name), length = names.length; i < length; i++) {
+        events = api(events, names[i], name[names[i]], context, ctx);
       }
-      return false;
+    } else if (name && eventSplitter.test(name)) {
+      // Handle space separated event names.
+      for (names = name.split(eventSplitter), length = names.length; i < length; i++) {
+        events = api(events, names[i], callback, context, ctx);
+      }
+    } else {
+      events = api(events, name, callback, context, ctx);
     }
+    return events;
+  };
+
+  // Handles actually adding the event handler.
+  var onApi = function(events, name, callback, context, ctx) {
+    if (callback) {
+      var handlers = events[name] || [];
+      events[name] = handlers.concat({callback: callback, context: context, ctx: context || ctx});
+    }
+    return events;
+  };
+
+  // Handles removing any event handlers that are no longer wanted.
+  var offApi = function(events, name, callback, context) {
+    // Remove all callbacks for all events.
+    if (!events || !name && !context && !callback) return;
+
+    var names = name ? [name] : _.keys(events);
+    for (var i = 0, length = names.length; i < length; i++) {
+      name = names[i];
+
+      // Bail out if there are no events stored.
+      var handlers = events[name];
+      if (!handlers) continue;
+
+      // Find any remaining events.
+      var remaining = [];
+      if (callback || context) {
+        for (var j = 0, k = handlers.length; j < k; j++) {
+          var handler = handlers[j];
+          if (
+            callback && callback !== handler.callback &&
+              callback !== handler.callback._callback ||
+                context && context !== handler.context
+          ) {
+            remaining.push(handler);
+          }
+        }
+      }
 
-    // Handle space separated event names.
-    if (eventSplitter.test(name)) {
-      var names = name.split(eventSplitter);
-      for (var i = 0, length = names.length; i < length; i++) {
-        obj[action].apply(obj, [names[i]].concat(rest));
+      // Replace events if there are any remaining.  Otherwise, clean up.
+      if (remaining.length) {
+        events[name] = remaining;
+      } else {
+        delete events[name];
       }
-      return false;
     }
+    return _.isEmpty(events) ? void 0 : events;
+  };
 
-    return true;
+  var triggerApi = function(obj, name, sentinel, args) {
+    if (obj._events) {
+      if (sentinel !== triggerApi) args = [sentinel].concat(args);
```

Comment:
That's because `trigger({ event: value}, 1, 2, 3)` will prepend `value` to the arguments of the listener. Re: https://github.com/jashkenas/backbone/issues/3460